cleverhans/attacks.py

```diff
@@ -1,4 +1,313 @@
+from abc import ABCMeta, abstractmethod
 import numpy as np
+import warnings
+
+from .utils import random_targets
+
+
+class Attack:
+    """
+    Abstract base class for all attack classes. All attacks must override the
+    `craft` method that returns adversarial examples corresponding to the input
+    data. However, the `generate_symbolic` method is optional.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, x, pred, back='tf', sess=None, clip_min=None,
+                 clip_max=None, params={}):
+        """
+        :param x: The model's symbolic inputs.
```

Comment:
We decided to put it in the constructor based on a recommendation from Alex. He said that if we create a new new symbolic graph every time we call generate(), it is possible that some backends would be consuming more and more memory each time. So eventually it may cause program to run out of memory.